File: crates/czdev/src/bump.rs
use anyhow::{anyhow, Context, Result};
use std::path::{Path, PathBuf};
use std::process::Command;

const PACKAGES_INDEX_URL: &str =
    "https://cardputerzero.github.io/packages/dists/stable/main/binary-arm64/Packages";

pub fn run(deb: Option<&Path>) -> Result<()> {
    let deb_path = resolve_deb(deb)?;

    let package = dpkg_field(&deb_path, "Package")?;
    let current_ver = dpkg_field(&deb_path, "Version")?;

    println!("Package: {package}");
    println!("Current version in deb: {current_ver}");

    // Fetch latest published version
    let latest = fetch_latest_version(&package)?;
    let next = match &latest {
        Some(v) => {
            println!("Latest published version: {v}");
            bump_patch(v)
        }
        None => {
            println!("No published version found (new package)");
            bump_patch(&current_ver)
        }
    };

    println!();
    println!("Next version: {next}");
    println!();
    println!("To rebuild with this version, update your package control file:");
    println!("  Version: {next}");
    println!();
    println!("Or if using CMake/packaging scripts, set:");
    println!("  PKG_VERSION={next}");

    Ok(())
}

fn resolve_deb(deb: Option<&Path>) -> Result<PathBuf> {
    if let Some(p) = deb {
        if !p.is_file() {
            return Err(anyhow!("file not found: {}", p.display()));
        }
        return Ok(p.to_path_buf());
    }
    let build_dir = Path::new("build");
    if build_dir.is_dir() {
        let mut debs: Vec<_> = std::fs::read_dir(build_dir)?
            .filter_map(|e| e.ok())
            .filter(|e| e.path().extension().map(|x| x == "deb").unwrap_or(false))
            .collect();
        if debs.len() == 1 {
            return Ok(debs.remove(0).path());
        }
        if debs.len() > 1 {
            return Err(anyhow!(
                "multiple .deb files in build/. Specify one with --deb <path>"
            ));
        }
    }
    Err(anyhow!("no .deb file found. Specify with --deb <path>"))
}

fn dpkg_field(deb: &Path, field: &str) -> Result<String> {
    let output = Command::new("dpkg-deb")
        .args(["-f", &deb.to_string_lossy(), field])
        .output()
        .with_context(|| format!("dpkg-deb -f {field}"))?;
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

fn fetch_latest_version(package: &str) -> Result<Option<String>> {
    let resp = reqwest::blocking::get(PACKAGES_INDEX_URL);
    let content = match resp {
        Ok(r) if r.status().is_success() => r.text().unwrap_or_default(),
        _ => return Ok(None),
    };

    let mut in_package = false;
    let mut latest: Option<String> = None;

    for line in content.lines() {
        if line.starts_with("Package: ") {
            in_package = line.trim_start_matches("Package: ") == package;
        }
        if in_package && line.starts_with("Version: ") {
            let ver = line.trim_start_matches("Version: ").to_string();
            match &latest {
                Some(existing) if compare_versions(&ver, existing) == std::cmp::Ordering::Greater => {
                    latest = Some(ver);
                }
                None => latest = Some(ver),
                _ => {}
            }
        }
        if line.is_empty() {
            in_package = false;
        }
    }
    Ok(latest)
}

fn bump_patch(version: &str) -> String {
    // Handle versions like "1.0.0", "1.0", "0.2.0-1~lofibox23"
    // Strip debian revision (everything after first '-')
    let base = version.split('-').next().unwrap_or(version);
    let parts: Vec<&str> = base.split('.').collect();

    match parts.len() {
        1 => format!("{}.0.1", parts[0]),
        2 => format!("{}.{}.1", parts[0], parts[1]),
        _ => {
            let patch: u64 = parts[2].parse().unwrap_or(0);
            format!("{}.{}.{}", parts[0], parts[1], patch + 1)
        }
    }
}

fn compare_versions(a: &str, b: &str) -> std::cmp::Ordering {
    let parse = |v: &str| -> Vec<u64> {
        v.split(|c: char| c == '.' || c == '-' || c == '~')
            .filter_map(|s| s.parse::<u64>().ok())
            .collect()
    };
    parse(a).cmp(&parse(b))
}

File: scripts/czdev/bump.py
"""Show next version (patch bump) for a package — mirrors the Rust bump module."""

import re
import subprocess
import sys
import urllib.request
from pathlib import Path
from typing import Optional

PACKAGES_INDEX_URL = "https://cardputerzero.github.io/packages/dists/stable/main/binary-arm64/Packages"


def run(deb: Optional[str] = None):
    deb_path = resolve_deb(deb)

    package = dpkg_field(deb_path, "Package")
    current_ver = dpkg_field(deb_path, "Version")

    print(f"Package: {package}")
    print(f"Current version in deb: {current_ver}")

    latest = fetch_latest_version(package)
    if latest:
        print(f"Latest published version: {latest}")
        next_ver = bump_patch(latest)
    else:
        print("No published version found (new package)")
        next_ver = bump_patch(current_ver)

    print()
    print(f"Next version: {next_ver}")
    print()
    print("To rebuild with this version, update your package control file:")
    print(f"  Version: {next_ver}")
    print()
    print("Or if using CMake/packaging scripts, set:")
    print(f"  PKG_VERSION={next_ver}")


def resolve_deb(deb: Optional[str]) -> str:
    if deb:
        if not Path(deb).is_file():
            print(f"file not found: {deb}", file=sys.stderr)
            sys.exit(1)
        return deb
    build_dir = Path("build")
    if build_dir.is_dir():
        debs = list(build_dir.glob("*.deb"))
        if len(debs) == 1:
            return str(debs[0])
        if len(debs) > 1:
            print("multiple .deb files in build/. Specify one with --deb <path>", file=sys.stderr)
            sys.exit(1)
    print("no .deb file found. Specify with --deb <path>", file=sys.stderr)
    sys.exit(1)


def dpkg_field(deb_path: str, field: str) -> str:
    try:
        result = subprocess.run(["dpkg-deb", "-f", deb_path, field],
                                capture_output=True, text=True, check=True)
        return result.stdout.strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        print(f"ERROR: dpkg-deb -f {field} failed", file=sys.stderr)
        sys.exit(1)


def fetch_latest_version(package: str) -> Optional[str]:
    try:
        resp = urllib.request.urlopen(PACKAGES_INDEX_URL, timeout=10)
        content = resp.read().decode()
    except Exception:
        return None

    in_package = False
    latest = None

    for line in content.splitlines():
        if line.startswith("Package: "):
            in_package = (line[len("Package: "):] == package)
        if in_package and line.startswith("Version: "):
            ver = line[len("Version: "):]
            if latest is None or compare_versions(ver, latest) > 0:
                latest = ver
        if line == "":
            in_package = False

    return latest


def bump_patch(version: str) -> str:
    base = version.split("-")[0]
    parts = base.split(".")
    if len(parts) == 1:
        return f"{parts[0]}.0.1"
    elif len(parts) == 2:
        return f"{parts[0]}.{parts[1]}.1"
    else:
        try:
            patch = int(parts[2])
        except ValueError:
            patch = 0
        return f"{parts[0]}.{parts[1]}.{patch + 1}"


def compare_versions(a: str, b: str) -> int:
    def parse(v):
        return [int(x) for x in re.split(r'[.\-~]', v) if x.isdigit()]
    pa, pb = parse(a), parse(b)
    if pa > pb:
        return 1
    elif pa < pb:
        return -1
    return 0
